mockserver-netty/src/main/java/org/mockserver/unification/PortUnificationHandler.java

```diff
@@ -185,22 +190,32 @@ private void switchToHttp(ChannelHandlerContext ctx, ByteBuf msg) {
     }
 
     private Set<String> getLocalAddresses(ChannelHandlerContext ctx) {
-        Set<String> localAddresses = new HashSet<>();
         SocketAddress localAddress = ctx.channel().localAddress();
-        if (localAddress instanceof InetSocketAddress) {
+        if (!(localAddress instanceof InetSocketAddress)) {
+            return new HashSet<>();
+        }
+
+        synchronized (localAddresses) {
```

Comment:
Please replace the synchronized HashMap with a ConcurrentHashMap, which has much more efficient locked and concurrency controls. In addition once you have done this there will be no need to wrap these lines in a synchronized block which would cause performance issues when under heavy load.

Reply:
Well, I'm not sure you correctly understand the intention.
If you want to protect iterations or query-then-insert (like done here), you have to lock the map.
With `ConcurrentHashMap` you cannot do this, as it locks on various individual buckets internally.
With `synchronizedMap` you have the explicit possibility to lock the map for iteration or query-then-insert operations.
With Java 8 I'd have used a `ConcurrentHashMap` and its `computeIfAbsent` method.
But with Java 7 this is not possible.

So to increase concurrency I could either change this to double-checked locking to only enter the `synchronized` block if actually a value has to be computed, or I can remove the `synchronized` block, then the same value gets computed multiple times when only once would have been necessary.

Waht do you prefer?

Reply:
The query then insertion in this case is a safe operation to be done because at most the worst case is you update the Map twice, as you pointed out.  You are not querying modifying and then inserting so in my opinion no need to add the additional synchronisation.  Especially given the value you are adding in immutable.  

I would suggest there at worst the same value being computed multiple times is acceptable, given as you correct discovered these operations can be slow and adding synchronisation around this could cause an issue because all thread will block and the Netty thread pool is limited.  If multiple threads block it is fairly likely the Netty thread pool will not be able to satisfy incoming requests and all requests processing will slow down.  This has happened before in MockServer for similar code and so is a real concern and not just a theoretical concern.

Do you agree?

Reply:
Sure.

As I said, the alternative would be double-checked locking. (in case you don't know the term, it would mean checking whether there is a value in the map before the synchronized block and returning it if it is alread present and only entering the synchronized block if there is actually a value to be computed to prevent it from being computed multiple times)

With that in most cases the synchronized block would not be entered, as the value will already be there and thus the threads would not block, except there is really the value calculated right now.

But I agree, as it is not about having the same value on each calculation, but to save time if there is already a value calculated it is probably not tragic if the value gets calculated multiple times if two requests really need it concurrently and the parallel requests will probably have to wait about the same amount of time. And also maybe because of the more efficient implementation there is even more time saved by using the CHM, while I expect this gain to be minimal actually.